### DISCUSSION TITLE
handling-web-api-exceptions-with-problemdetails-middleware/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#9] [Announcements]
# handling-web-api-exceptions-with-problemdetails-middleware/

In this post I describe the ProblemDetailsMiddleware by Kristian Hellang that can be used to return ProblemDetails when an exception occurs in an API project.

https://andrewlock.net/handling-web-api-exceptions-with-problemdetails-middleware/

**Comment by AKlaus** [comment]
Hi Andrew,
Came across this post and thought you (or future readers) might be interested to check out an alternative solution to the problem of converting a successful/failed outcome of the Domain logic to a meaningful `ProblemDetails` in the API response.

It's [DomainResult](https://github.com/AKlaus/DomainResult) NuGet package that instead of mapping exceptions to problem details, introduces a wrapper around the return type in the Domain method and a set of extensions that convert the wrapped output to the corresponding HTTP 2xx code or 4xx/5xx along with a `ProblemDetails`.

The difference between `Hellang.Middleware.ProblemDetails` and `ProblemDetails` is ideological and most visible in the sad path of execution – "_throwing an exception_" vs "_returning a faulty status_".

The main distinctive features of `DomainResult` are

 - Allows simpler nested calls of the domain logic (no exceptions handlers when the severity of their "sad" path is not exception-worthy).
 - Provides a predefined set of responses for main execution paths ("bad request", "not found", etc.). Works out-of-the-box.
 - Has an option to tune the conversion to the HTTP response for each output independently.
 - Bonus – auto handling of HTTP code 204 (if no response value) and gives a simple way to form a proper HTTP 201 response.

I'd be glad if it helps someone,
Alex

P.S. Full disclosure – I'm the creator of `DomainResult`.

**Comment by zwoolli** [comment]
In Net7.0+ the `IServiceCollection` has [`ProblemDetailsServiceCollectionExtensions`](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/src/ProblemDetailsServiceCollectionExtensions.cs), which have an `AddProblemDetails` extension. This makes the call to Hellang's `AddProblemDetails` extension ambiguous. How do you get around this?

[a comment by an automated account is not shown]
